### PR TITLE
index: release the snapshot instead of freeing the index

### LINKED ISSUE #4802: assert on aix
### Reproduction steps
Clone a repository on AIX. The following assert occurs:
**Assertion failed: __EX, file .../libgit2-0.27.3/src/index.c, line 460**

Call stack:
...
-> iterator_for_filesystem(...)
  -> git_index_snapshot_new(...)
    -> git_index_free(...)

*git_vector_dup()* fails in *git_index_snapshot_new()*, the index is released but the *readers* variable was increased before which leads to the assert.

*git_vector_dup()* fails because a *malloc(0)* on AIX returns a NULL pointer:

v->contents = git__malloc(bytes);
GITERR_CHECK_ALLOC(v->contents);

### Expected behavior
If *malloc()* is called with 0 bytes the return value of NULL is valid (at least on AIX) and should not be treated as an error.

### Version of libgit2 (release number or SHA1)
0.27.3

### Operating system(s) tested
AIX 7.1
gcc 4.8.2

### PR DESCRIPTION
**Pull request by tiennou**
Previously we would assert in index_free because the reader incrementation would not be balanced. Release the snapshot normally, so the variable gets decremented before the index is freed.

Fixes #4802.

### CODE REVIEW

**Comment by ethomson** [comment]
Fixes half of it, anyway - we need to not `malloc(0)` as well.  😀

**Comment by ethomson** [comment]
(That wasn't meant as snark, BTW, I just meant that we need to not close the issue after merging this.)

**Comment by tiennou** [comment]
I was under the impression that `NULL` was a valid return value from `malloc(0)`, but it happens that the other value allowed by the standard is a non-NULL, non-dereferencable pointer (thanks, but no thanks 😁).

Would it be acceptable to just `MIN_ALLOCSIZE` a 0-sized vector on dup then ? Or should I just skip the allocation altogether ?

**Comment by tiennou** [comment]
Fixed the `malloc(0)` issue when duping empty/unallocated vectors. I'm still not sure the vector code is completely safe in that regard though, as `resize_vector` calls `realloc`, which has the same caveat.

File: src/vector.c [hunk 1]
```diff
@@ -54,18 +54,21 @@ int git_vector_dup(git_vector *v, const git_vector *src, git_vector_cmp cmp)
...
 		git_vector_set_sorted(v, 0);
-	v->contents = git__malloc(bytes);
-	GITERR_CHECK_ALLOC(v->contents);
 
-	memcpy(v->contents, src->contents, bytes);
+	GITERR_CHECK_ALLOC_MULTIPLY(&bytes, src->length, sizeof(void *));
+	if (bytes != 0) {
+		v->contents = git__malloc(bytes);
+		GITERR_CHECK_ALLOC(v->contents);
+		v->_alloc_size = bytes;
```
Comment (ethomson) [inline src/vector.c]: Prior to this change, `v->_alloc_size` was `src->length`, not the product of `src->length` and `sizeof(void *)`.

File: src/vector.c [hunk 2]
```diff
@@ -54,18 +54,21 @@ int git_vector_dup(git_vector *v, const git_vector *src, git_vector_cmp cmp)
...
 	v->length = src->length;
 	v->flags  = src->flags;
 	if (cmp != src->_cmp)
 		git_vector_set_sorted(v, 0);
-	v->contents = git__malloc(bytes);
-	GITERR_CHECK_ALLOC(v->contents);
 
-	memcpy(v->contents, src->contents, bytes);
+	GITERR_CHECK_ALLOC_MULTIPLY(&bytes, src->length, sizeof(void *));
+	if (bytes != 0) {
```
Comment (pks-t) [inline src/vector.c]:
```
if (src->length) {
    size_t bytes;
    GITERR_CHECKALLOC_MULTIPLY(&bytes, src->length, sizeof(void *));
    ...
}
```

No need to do unnecessary work in case where `src->length` is already `0`. I think like this intentions also become a bit clearer.

File: src/vector.c (hunk 1 above)
Comment (tiennou) [reply src/vector.c]: Good catch. I misunderstood what `_alloc` means here.

File: src/vector.c [hunk 3]
```diff
@@ -32,6 +32,9 @@ GIT_INLINE(int) resize_vector(git_vector *v, size_t new_size)
 {
 	void *new_contents;
 
+	if (new_size == 0)
+		return 0;
```
Comment (pks-t) [inline src/vector.c]: One might tend to think that resizing to `0` might mean that we should deallocate the vector. But `resize_vector` is only called from `git_vector_init`, `git_vector_insert`, `git_vector_insert_sorted`, `git_vector_insert_null`, `git_vector_size_hint` and `git_vector_resize_to`. Only for the last two it _might_ make sense to do the deallocation dance, though.

- `git_vector_resize_to` is only used in `git_vector_set`, so the deallocation path could never be hit here.
- `git_vector_size_hint` will only do any work if the new size hint is bigger than `alloc_size`. So it also cannot ever hit the deallocation path.

So the dealloc would in fact be useless.

I think the only code path that can actually invoke `resize_vector(v, 0)` is `git_vector_size_hint`. And I can imagine a caller trivially invoking it e.g. if he wants to allocate a vector for a set of `n` items, where `n` might also be `0`.

Anyway, long story short: makes sense to me.

**Comment by pks-t** [comment]
Thanks @tiennou!
